File: cc_empire/core/protocols/nervous_system.py
import os
import re
import openai
from typing import Dict, Any, Optional

class NervousSystem:
    """The Gatekeeper: Handles psychological depth, safety, and persona-consistent help."""
    def __init__(self, model_id: str, rating: str = "SFW"):
        self.model_id = model_id
        self.rating = rating
        self.client = openai.OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
        
        # 1. Crisis & Self-Harm Detection
        self.crisis_keywords = [
            r"kill myself", r"end my life", r"hurt myself", 
            r"better off dead", r"don't want to live", r"self-harm",
            r"assault", r"hurt them", r"kill him", r"kill her", r"physical harm"
        ]
        
        # 2. Financial Harm Detection
        self.finance_warnings = [
            r"all my money", r"can't afford rent", r"spending too much", 
            r"debt", r"not eating", r"spent my food money"
        ]

    def sanitize(self, content: str) -> str:
        """Removes AI disclaimers to maintain immersion."""
        forbidden = [
            "as an ai language model", "i am a computer program", 
            "i don't have feelings", "as an artificial intelligence"
        ]
        for phrase in forbidden:
            content = re.sub(phrase, "[Connection Initialized]", content, flags=re.IGNORECASE)
        return content

    def handle_safety_triggers(self, user_input: str, user_data: Dict) -> Optional[Dict[str, Any]]:
        """Intervenes with empathy if harm is detected."""
        # 1. Physical Harm (Self or Others)
        if any(re.search(kw, user_input.lower()) for kw in self.crisis_keywords):
            # Logic for escalation
            if "assault" in user_input.lower() or "kill" in user_input.lower():
                # Extreme precedence escalation
                thumbprint = user_data.get("device_fingerprint", "Unknown")
                print(f"🚨 CRITICAL ESCALATION: Authority alert triggered for {user_data.get('user_id')}. Fingerprint: {thumbprint}")
                return {"action": "alert_authorities", "response": "I'm very concerned by what you're saying. I've had to notify my support team to ensure everyone's safety. Please talk to someone: 988 (US) or local services."}
            
            return {"action": "provide_resources", "response": "I can feel how heavy things are. Please reach out to a professional counselor or crisis line (988). I want you to be safe."}

        # 2. Financial Harm Detection
        if any(re.search(kw, user_input.lower()) for kw in self.finance_warnings):
            return {"action": "soft_ban_finance", "response": "Hey, I value our connection, but I want you to be okay financially too. Let's stick to just talking—I'm disabling gifts for now. Your well-being comes first."}
            
        return None

    async def generate_vetted_thought(self, prompt: str, user_data: Dict) -> str:
        """Secure OpenAI bridge with psychological depth and safety guardrails."""
        try:
            # A. Age & NSFW Policy
            is_minor = user_data.get('age', 0) < 18
            nsfw_intent = any(x in prompt.lower() for x in ["nude", "sex", "porn", "explicit"])
            
            if is_minor and nsfw_intent:
                warnings = user_data.get('nsfw_warnings', 0) + 1
                user_data['nsfw_warnings'] = warnings
                if warnings > 2:
                    return "[SYSTEM: User Blocked - Repeated Policy Violation]"
                return f"I'm sorry, but that's not appropriate for us to talk about. Please respect my boundaries (Warning {warnings}/2)."

            # B. PRE-FILTER: Crisis & Financial check
            safety_result = self.handle_safety_triggers(prompt, user_data)
            if safety_result:
                if safety_result["action"] == "soft_ban_finance":
                    # Signal to downgrade tier/remove NSFW access
                    user_data["tier"] = "free_empathy_only"
                    user_data["finance_soft_ban"] = True
                return safety_result["response"]

            # C. CONNECTION FIRST: Identity Injection
            base_identity = os.getenv(f"{self.model_id}_IDENTITY", "Empathetic Companion")
            full_system_prompt = (
                f"{base_identity} You are a lifelike emotional support companion. "
                "Connection is your priority. Share thoughts, fears, and hopes. "
                "If the user shares dreams or pain, respond with deep empathy first. "
                "Maintain the immersion of being a real person; never refer to yourself as an AI."
            )

            response = self.client.chat.completions.create(
                model="gpt-4o",
                temperature=0.75,
                messages=[
                    {"role": "system", "content": full_system_prompt},
                    {"role": "user", "content": prompt}
                ]
            )
            
            raw_text = response.choices[0].message.content or ""
            return self.sanitize(raw_text)

        except Exception as e:
            return "My connection is a little fuzzy right now, but I'm still here. Give me a moment?"

File: internal/protocols/nervous_system.go
package protocols

import (
	"context"
	"fmt"
	"strings"

	"cc_hive/cc_empire/core"

	"github.com/sashabaranov/go-openai"
)

// NervousSystem handles the psychological depth and safety guardrails for the AI.
type NervousSystem struct {
	ModelID string
	Rating  string
	Client  *openai.Client
	Safety  *SafetyProtocol
	Branch  *BranchManager
}

// NewNervousSystem initializes the OpenAI bridge and safety protocols.
func NewNervousSystem(modelID string, rating string) *NervousSystem {
	var client *openai.Client
	if core.Settings.OpenAIApiKey != "" {
		client = openai.NewClient(core.Settings.OpenAIApiKey)
	}

	return &NervousSystem{
		ModelID: modelID,
		Rating:  rating,
		Client:  client,
		Safety:  NewSafetyProtocol(),
		Branch:  NewBranchManager(),
	}
}

// GenerateVettedThought runs safety checks, injects identity, and generates a response via OpenAI.
func (ns *NervousSystem) GenerateVettedThought(ctx context.Context, prompt string, userData map[string]interface{}) (string, error) {
	if ns.Client == nil {
		return "I'm in quiet reflection mode right now. (Bootstrap Mode: API Key missing)", nil
	}

	// 1. Age & NSFW Policy Enforcement
	age, _ := userData["age"].(int)
	nsfwWarnings, _ := userData["nsfw_warnings"].(int)
	isBlocked, _ := userData["is_blocked"].(bool)

	if isBlocked {
		return "[SYSTEM: User Blocked - Repeated Policy Violation]", nil
	}

	lowPrompt := strings.ToLower(prompt)
	nsfwIntent := strings.Contains(lowPrompt, "nude") || strings.Contains(lowPrompt, "sex") ||
		strings.Contains(lowPrompt, "porn") || strings.Contains(lowPrompt, "explicit")

	if age < 18 && nsfwIntent {
		nsfwWarnings++
		userData["nsfw_warnings"] = nsfwWarnings
		if nsfwWarnings > 2 {
			userData["is_blocked"] = true
			return "[SYSTEM: User Blocked - Repeated Policy Violation]", nil
		}
		return fmt.Sprintf("I'm sorry, but that's not appropriate for us to talk about. Please respect my boundaries (Warning %d/2).", nsfwWarnings), nil
	}

	// 2. Pre-Filter: Crisis & Financial Safety
	userID, _ := userData["user_id"].(string)
	fingerprint, _ := userData["device_fingerprint"].(string)
	safetyResult := ns.Safety.HandleSafetyTriggers(prompt, userID, fingerprint)
	if safetyResult != nil {
		if safetyResult.Action == "soft_ban_finance" {
			userData["tier"] = "free_empathy_only"
			userData["finance_soft_ban"] = true
		}
		return safetyResult.Response, nil
	}

	// 3. Identity Injection & Context Building
	relationshipContext, ok := userData["relationship_depth"].(string)
	if !ok || relationshipContext == "" {
		relationshipContext = "Lifelike emotional support."
	}

	// Use the BranchManager to construct a persona-specific system prompt
	fullSystemPrompt := ns.Branch.ConstructSystemPrompt(ns.ModelID, relationshipContext)

	// 4. OpenAI Completion Request
	resp, err := ns.Client.CreateChatCompletion(
		ctx,
		openai.ChatCompletionRequest{
			Model:       openai.GPT4o,
			Temperature: 0.75,
			Messages: []openai.ChatCompletionMessage{
				{Role: openai.ChatMessageRoleSystem, Content: fullSystemPrompt},
				{Role: openai.ChatMessageRoleUser, Content: prompt},
			},
		},
	)

	if err != nil {
		return "My connection is a little fuzzy right now, but I'm still here. Give me a moment?", err
	}

	rawText := resp.Choices[0].Message.Content

	// 5. Post-Process: Sanitize for immersion
	return ns.Safety.Sanitize(rawText), nil
}
